Add filtering to endpoint that provides DataElements per DataModel

jamesrwelch opened the pull request:
One failing test which I think is unrelated

Resolves #470

jamesrwelch committed:
Potential fix for #470

pjmonks commented on mdm-plugin-datamodel/grails-app/services/uk/ac/ox/softeng/maurodatamapper/datamodel/item/DataElementService.groovy:
```diff
@@ -310,8 +310,26 @@ WHERE (de.dataClass.id = :dataClassId OR idc.id = :dataClassId)''', 'de', filter
-    List<DataElement> findAllByDataModelId(Serializable dataModelId, Map pagination = [:]) {
-        DataElement.byDataModelId(dataModelId).list(pagination)
+    List<DataElement> findAllByDataModelId(Serializable dataModelId, Map filters = [:], Map pagination = [:]) {
+        Map<String, Object> queryParams = [dataModelId: dataModelId]
```
I've spotted a problem when testing this locally. I tried `GET /dataModels/{dataModelId}/dataElements?max=20&label=accomodation` to test filtering the data element label but got a 500 response saying "Named parameter not bound : label" from Hibernate.

I compared to the data type list endpoint `GET /dataModels/{dataModelId}/dataTypes?max=20&label=abdom` which works. Comparing the `DataTypeService`, it has this line in `findAllDataTypesByHQLQuery()`:

```groovy
queryParams.putAll(extractFiltersAsHQLParameters(filters))
```

This isn't included in the `DataElementService`. So I think it's still only filtering on `dataModelId` but the HQL query now includes a `:label` parameter it cannot resolve.

pjmonks reviewed (changes requested):
Seems close, but returns an error for me, please review my comment.

jamesrwelch committed:
Fix to parameter interpretation

jamesrwelch commented:
@OButlerOcc @pjmonks  I believe this is now ready for re-review.  Note the force-push after rebasing on develop.

pjmonks committed:
gh-470 Update list all data elements by data model filters

- Support filtering by label of parent data class

pjmonks reviewed (approved):
I've tested your fix again and it works now. I've also pushed up my own addition, I figured out how to filter by both label for the data element and the parent data class too. If any tests fail on Jenkins I believe they will be unrelated. Ready to merge ✔️

joe-crawford commented:
@pjmonks, @jamesrwelch FYI I've tried running the failing tests locally, `FolderTreeItemFunctionalSpec` passed but `VersionedFolderFunctionalSpec` failed at the same test (DI06) as in Jenkins.

joe-crawford commented:
VersionedFolderFunctionalSpec DI06 is passing locally on develop so think something has been affected there. Not sure exactly what though.

pjmonks committed:
gh-470 Fix failing test that found the results were not unwrapping all proxy objects

VersionedFolderFunctionalSpec test 'DI06 : test diffing 2 complex versioned folders (as reader of both models)' was failing

Reason was that the DataElementService.findAllByDataModelId() was being called and the breadcrumb tree of each result was a Hibernate proxy which broken the view rendering of the diff

pjmonks committed:
gh-470 Fix failing tests for rendering data elements

- Add tests to verify that new filter works correctly on "List all data elements for data model" endpoint
- Fix issue rendering result set by unwrapping Hibernate proxy objects
